### PR TITLE
Fix new warnings generated by GCC 7

### LINKED ISSUE #1375: GCC 7.1.1 reporting lots of new warnings
Apparently my Arch VM updated to a new GCC version, I noticed a whole slew of warnings coming through, thought I would try to document the collection here.  It can be split up if that would be helpful.  Many of these are of the same new warning class.

```
[ 13%] Building CXX object code/CMakeFiles/code.dir/ai/aicode.cpp.o
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp: In function ‘void ai_get_dock_goal_indexes(object*, ai_info*, ai_goal*, object*, int&, int&)’:
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10397:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
    Warning(LOCATION, "Normally dock indexes should be calculated for only AIS_DOCK_0 and AIS_UNDOCK_0.  Trace out and debug.");
    ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10398:3: note: here
   case AIS_DOCK_0:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10414:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
    Warning(LOCATION, "Normally dock indexes should be calculated for only AIS_DOCK_0 and AIS_UNDOCK_0.  Trace out and debug.");
    ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10415:3: note: here
   case AIS_UNDOCK_0:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp: In function ‘void ai_cleanup_dock_mode_subjective(object*)’:
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10480:26: warning: this statement may fall through [-Wimplicit-fallthrough=]
     model_anim_start_type(goal_shipp, TRIGGER_TYPE_DOCKED, dockee_index, -1);
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10481:4: note: here
    case AIS_UNDOCK_1:
    ^~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10483:26: warning: this statement may fall through [-Wimplicit-fallthrough=]
     model_anim_start_type(goal_shipp, TRIGGER_TYPE_DOCKING_STAGE_3, dockee_index, -1);
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10484:4: note: here
    case AIS_UNDOCK_2:
    ^~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10492:26: warning: this statement may fall through [-Wimplicit-fallthrough=]
     model_anim_start_type(goal_shipp, TRIGGER_TYPE_DOCKED, dockee_index, -1);
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10493:4: note: here
    case AIS_DOCK_3:
    ^~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10495:26: warning: this statement may fall through [-Wimplicit-fallthrough=]
     model_anim_start_type(goal_shipp, TRIGGER_TYPE_DOCKING_STAGE_3, dockee_index, -1);
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:10496:4: note: here
    case AIS_DOCK_2:
    ^~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp: In function ‘void ai_maybe_evade_locked_missile(object*, ai_info*)’:
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:12377:6: warning: this statement may fall through [-Wimplicit-fallthrough=]
      if (object_is_docked(objp) || (aip->ai_flags[AI::AI_Flags::Repairing, AI::AI_Flags::Being_repaired]))
      ^~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:12379:5: note: here
     case AIM_GUARD:
     ^~~~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:12381:6: warning: this statement may fall through [-Wimplicit-fallthrough=]
      if ((aip->guard_objnum != -1) && (aip->guard_signature == Objects[aip->guard_objnum].signature)) {
      ^~
/usr/share/games/fs2source/fs2_open/code/ai/aicode.cpp:12386:5: note: here
     case AIM_EVADE:
     ^~~~
```

```
[ 15%] Building CXX object code/CMakeFiles/code.dir/cfile/cfilesystem.cpp.o
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp: In function ‘int cf_get_file_list(SCP_vector<std::__cxx11::basic_string<char> >&, int, const char*, int, SCP_vector<file_list_info>*)’:
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp:1489:5: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 253 [-Wformat-truncation=]
 int cf_get_file_list( SCP_vector<SCP_string> &list, int pathtype, const char *filter, int sort, SCP_vector<file_list_info> *info )
     ^~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp:1570:12: note: ‘snprintf’ output 2 or more bytes (assuming 257) into a destination of size 254
    snprintf(fn, MAX_PATH-1, "%s/%s", filespec, dir->d_name);
    ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp: In function ‘int cf_get_file_list(int, char**, int, const char*, int, file_list_info*)’:
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp:1695:5: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 253 [-Wformat-truncation=]
 int cf_get_file_list( int max, char **list, int pathtype, const char *filter, int sort, file_list_info *info )
     ^~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp:1779:12: note: ‘snprintf’ output 2 or more bytes (assuming 257) into a destination of size 254
    snprintf(fn, MAX_PATH-1, "%s/%s", filespec, dir->d_name);
    ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp: In function ‘int cf_get_file_list_preallocated(int, char (*)[32], char**, int, const char*, int, file_list_info*)’:
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp:1908:5: warning: ‘%s’ directive output may be truncated writing up to 255 bytes into a region of size 253 [-Wformat-truncation=]
 int cf_get_file_list_preallocated( int max, char arr[][MAX_FILENAME_LEN], char **list, int pathtype, const char *filter, int sort, file_list_info *info )
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/cfile/cfilesystem.cpp:1994:12: note: ‘snprintf’ output 2 or more bytes (assuming 257) into a destination of size 254
    snprintf(fn, MAX_PATH-1, "%s/%s", filespec, dir->d_name);
    ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
[ 24%] Building CXX object code/CMakeFiles/code.dir/graphics/opengl/gropengldraw.cpp.o
/usr/share/games/fs2source/fs2_open/code/graphics/opengl/gropengldraw.cpp: In function ‘void gr_opengl_deferred_lighting_finish()’:
/usr/share/games/fs2source/fs2_open/code/graphics/opengl/gropengldraw.cpp:2153:51: warning: this statement may fall through [-Wimplicit-fallthrough=]
     Current_shader->program->Uniforms.setUniform3f( "coneDir", l->vec2.xyz.x, l->vec2.xyz.y, l->vec2.xyz.z);
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/opengl/gropengldraw.cpp:2154:4: note: here
    case LT_POINT:
    ^~~~
```

```
[ 24%] Building CXX object code/CMakeFiles/code.dir/graphics/opengl/gropengldraw.cpp.o
/usr/share/games/fs2source/fs2_open/code/graphics/opengl/gropengldraw.cpp: In function ‘void gr_opengl_deferred_lighting_finish()’:
/usr/share/games/fs2source/fs2_open/code/graphics/opengl/gropengldraw.cpp:2153:51: warning: this statement may fall through [-Wimplicit-fallthrough=]
     Current_shader->program->Uniforms.setUniform3f( "coneDir", l->vec2.xyz.x, l->vec2.xyz.y, l->vec2.xyz.z);
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/opengl/gropengldraw.cpp:2154:4: note: here
    case LT_POINT:
    ^~~~
```

```
[ 26%] Building C object code/CMakeFiles/code.dir/graphics/paths/nanovg/nanovg.c.o
In file included from /usr/share/games/fs2source/fs2_open/code/graphics/paths/nanovg/nanovg.c:32:0:
/usr/share/games/fs2source/fs2_open/code/graphics/paths/nanovg/stb_image.h: In function ‘stbi__tga_get_comp’:
/usr/share/games/fs2source/fs2_open/code/graphics/paths/nanovg/stb_image.h:5230:18: warning: this statement may fall through [-Wimplicit-fallthrough=]
       case 16: if(is_grey) return STBI_grey_alpha;
                  ^
/usr/share/games/fs2source/fs2_open/code/graphics/paths/nanovg/stb_image.h:5232:7: note: here
       case 15: if(is_rgb16) *is_rgb16 = 1;
       ^~~~
[ 26%] Building CXX object code/CMakeFiles/code.dir/graphics/generic.cpp.o
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp: In function ‘int generic_anim_stream(generic_anim*, bool)’:
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:157:5: warning: ‘_0000’ directive output may be truncated writing 5 bytes into a region of size between 1 and 32 [-Wformat-truncation=]
 int generic_anim_stream(generic_anim *ga, const bool cache)
     ^~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:254:11: note: ‘snprintf’ output between 6 and 37 bytes into a destination of size 32
   snprintf(frame_name, MAX_FILENAME_LEN, "%s_0000", ga->filename);
   ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:157:5: warning: ‘_0001’ directive output may be truncated writing 5 bytes into a region of size between 1 and 32 [-Wformat-truncation=]
 int generic_anim_stream(generic_anim *ga, const bool cache)
     ^~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:260:11: note: ‘snprintf’ output between 6 and 37 bytes into a destination of size 32
   snprintf(frame_name, MAX_FILENAME_LEN, "%s_0001", ga->filename);
   ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp: In function ‘void generic_render_eff_stream(generic_anim*)’:
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:356:6: warning: ‘%.4d’ directive output may be truncated writing between 4 and 11 bytes into a region of size between 0 and 31 [-Wformat-truncation=]
 void generic_render_eff_stream(generic_anim *ga)
      ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:372:11: note: ‘snprintf’ output between 6 and 44 bytes into a destination of size 32
   snprintf(frame_name, MAX_FILENAME_LEN, "%s_%.4d", ga->filename, ga->current_frame);
   ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:356:6: warning: ‘_0001’ directive output may be truncated writing 5 bytes into a region of size between 1 and 32 [-Wformat-truncation=]
 void generic_render_eff_stream(generic_anim *ga)
      ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/graphics/generic.cpp:382:13: note: ‘snprintf’ output between 6 and 37 bytes into a destination of size 32
     snprintf(frame_name, MAX_FILENAME_LEN, "%s_0001", ga->filename);
     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
[ 28%] Building CXX object code/CMakeFiles/code.dir/hud/hudartillery.cpp.o
/usr/share/games/fs2source/fs2_open/code/hud/hudartillery.cpp: In function ‘void parse_ssm(const char*)’:
/usr/share/games/fs2source/fs2_open/code/hud/hudartillery.cpp:146:21: warning: this statement may fall through [-Wimplicit-fallthrough=]
      required_string("Circle");
      ~~~~~~~~~~~~~~~^~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/hud/hudartillery.cpp:147:5: note: here
     case -1: // If we're ignoring parse errors and can't identify the shape, go with a circle.
     ^~~~
```

```
[ 31%] Building CXX object code/CMakeFiles/code.dir/inetfile/cftp.cpp.o
/usr/share/games/fs2source/fs2_open/code/inetfile/cftp.cpp: In member function ‘uint CFtpGet::GetFile()’:
/usr/share/games/fs2source/fs2_open/code/inetfile/cftp.cpp:255:6: warning: ‘%s’ directive writing up to 199 bytes into a region of size 196 [-Wformat-overflow=]
 uint CFtpGet::GetFile()
      ^~~~~~~
/usr/share/games/fs2source/fs2_open/code/inetfile/cftp.cpp:272:10: note: ‘sprintf’ output between 7 and 206 bytes into a destination of size 200
   sprintf(szCommandString,"CWD %s\r\n",m_szDir);
   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
[ 37%] Building CXX object code/CMakeFiles/code.dir/menuui/credits.cpp.o
/usr/share/games/fs2source/fs2_open/code/menuui/credits.cpp: In function ‘void credits_do_frame(float)’:
/usr/share/games/fs2source/fs2_open/code/menuui/credits.cpp:690:3: warning: this statement may fall through [-Wimplicit-fallthrough=]
   if ( !(Player->flags & PLAYER_FLAGS_IS_MULTI) ) {
   ^~
/usr/share/games/fs2source/fs2_open/code/menuui/credits.cpp:696:2: note: here
  case KEY_CTRLED | KEY_DOWN:
  ^~~~
```

```
[ 40%] Building CXX object code/CMakeFiles/code.dir/mission/missionhotkey.cpp.o
/usr/share/games/fs2source/fs2_open/code/mission/missionhotkey.cpp: In function ‘void hotkey_button_pressed(int)’:
/usr/share/games/fs2source/fs2_open/code/mission/missionhotkey.cpp:881:16: warning: this statement may fall through [-Wimplicit-fallthrough=]
    save_hotkeys();
    ~~~~~~~~~~~~^~
/usr/share/games/fs2source/fs2_open/code/mission/missionhotkey.cpp:884:3: note: here
   case CANCEL_BUTTON:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/mission/missionhotkey.cpp: In function ‘void mission_hotkey_do_frame(float)’:
/usr/share/games/fs2source/fs2_open/code/mission/missionhotkey.cpp:1053:16: warning: this statement may fall through [-Wimplicit-fallthrough=]
    save_hotkeys();
    ~~~~~~~~~~~~^~
/usr/share/games/fs2source/fs2_open/code/mission/missionhotkey.cpp:1056:3: note: here
   case KEY_ESC:
   ^~~~
```

```
[46%] Building CXX object code/CMakeFiles/code.dir/model/modelrender.cpp.o
/usr/share/games/fs2source/fs2_open/code/model/modelrender.cpp: In function ‘void model_render_glowpoint(int, vec3d*, matrix*, glow_point_bank*, glow_point_bank_override*, polymodel*, ship*, bool)’:
/usr/share/games/fs2source/fs2_open/code/model/modelrender.cpp:1741:6: warning: this statement may fall through [-Wimplicit-fallthrough=]
      if((gpo && gpo->off_time_override)?gpo->off_time:bank->off_time) {
      ^~
/usr/share/games/fs2source/fs2_open/code/model/modelrender.cpp:1747:5: note: here
     case PULSE_TRI:
     ^~~~
```

```
[ 49%] Building CXX object code/CMakeFiles/code.dir/network/multi_pxo.cpp.o
/usr/share/games/fs2source/fs2_open/code/network/multi_pxo.cpp: In function ‘int multi_pxo_connect_do()’:
/usr/share/games/fs2source/fs2_open/code/network/multi_pxo.cpp:1813:5: warning: ‘:’ directive writing 1 byte into a region of size between 0 and 255 [-Wformat-overflow=]
 int multi_pxo_connect_do()
     ^~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/network/multi_pxo.cpp:1830:10: note: ‘sprintf’ output between 6 and 261 bytes into a destination of size 255
   sprintf(ip_string, "%s:%d", Multi_options_g.pxo_ip, PXO_CHAT_PORT);
   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
[ 58%] Building CXX object code/CMakeFiles/code.dir/parse/sexp.cpp.o
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp: In function ‘void sexp_manipulate_colgroup(int, bool)’:
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:22297:23: warning: ‘<<’ in boolean context, did you mean ‘<’ ? [-Wint-in-bool-context]
     colgroup_id &= !(1<<group);
                     ~~^~~~~~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp: In function ‘int sexp_is_iff(int)’:
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:9250:5: warning: this statement may fall through [-Wimplicit-fallthrough=]
     }
     ^
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:9254:4: note: here
    default:
    ^~~~~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp: In function ‘void sexp_alter_ship_flag_helper(object_ship_wing_point_team&, bool, Object::Object_Flags, Ship::Ship_Flags, Mission::Parse_Object_Flags, AI::AI_Flags, bool)’:
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:13017:4: warning: this statement may fall through [-Wimplicit-fallthrough=]
    if (ai_flag != AI::AI_Flags::NUM_VALUES)
    ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:13025:3: note: here
   case OSWPT_TYPE_PARSE_OBJECT:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp: In function ‘int query_operator_argument_type(int, int)’:
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26151:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
    } else if (argnum == 2) {
           ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26155:3: note: here
   case OP_INT_TO_STRING:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26158:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
    } else if (argnum == 1) {
           ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26162:3: note: here
   case OP_STRING_GET_SUBSTRING:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26167:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
    } else if (argnum == 3) {
           ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26171:3: note: here
   case OP_STRING_SET_SUBSTRING:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26178:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
    } else if (argnum == 4) {
           ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26182:3: note: here
   case OP_DEBUG:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26588:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum == 1)
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26591:3: note: here
   case OP_DISTANCE_SUBSYSTEM:
   ^~~~
In file included from /usr/share/games/fs2source/fs2_open/code/ai/aigoals.h:16:0,
                 from /usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:25:
/usr/share/games/fs2source/fs2_open/code/globalincs/pstypes.h:280:27: warning: this statement may fall through [-Wimplicit-fallthrough=]
  #define Int3() debug_int3(__FILE__, __LINE__)
                 ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26599:5: note: in expansion of macro ‘Int3’
     Int3();  // shouldn't happen
     ^
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26601:3: note: here
   case OP_NUM_WITHIN_BOX:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26640:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum == 1)
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26643:3: note: here
   case OP_IS_SUBSYSTEM_DESTROYED_DELAY:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26870:10: warning: this statement may fall through [-Wimplicit-fallthrough=]
    argnum--;
    ~~~~~~^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26872:3: note: here
   case OP_UPDATE_SOUND_ENVIRONMENT:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26973:24: warning: this statement may fall through [-Wimplicit-fallthrough=]
     return OPF_POSITIVE;
                        ^
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:26976:3: note: here
   case OP_TRAINING_MSG:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27118:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum == 2)
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27121:3: note: here
   case OP_GOAL_INCOMPLETE:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27428:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
    } else if(argnum > 2) {
           ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27432:3: note: here
   case OP_TURRET_SET_DIRECTION_PREFERENCE:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27562:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum == 5)
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27565:3: note: here
   case OP_BEAM_FREE_ALL:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27991:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum == 12 )
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:27994:3: note: here
   case OP_CUTSCENES_SHOW_SUBTITLE_TEXT:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:28005:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum == 12)
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:28008:3: note: here
   case OP_CUTSCENES_SHOW_SUBTITLE_IMAGE:
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:28017:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum == 9)
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:28022:3: note: here
   case OP_JUMP_NODE_SET_JUMPNODE_NAME: //CommanderDJ
   ^~~~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:28025:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
    else if (argnum==1)
         ^~
/usr/share/games/fs2source/fs2_open/code/parse/sexp.cpp:28028:3: note: here
   case OP_JUMP_NODE_SET_JUMPNODE_COLOR:
   ^~~~
```

```
[ 65%] Building CXX object code/CMakeFiles/code.dir/scripting/ade.cpp.o
/usr/share/games/fs2source/fs2_open/code/scripting/ade.cpp: In member function ‘void scripting::ade_table_entry::OutputMeta(FILE*)’:
/usr/share/games/fs2source/fs2_open/code/scripting/ade.cpp:681:12: warning: this statement may fall through [-Wimplicit-fallthrough=]
     Warning(LOCATION, "Unknown type '%c' passed to ade_table_entry::OutputMeta", Type);
     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/scripting/ade.cpp:682:4: note: here
    case 'b':
    ^~~~
```

```
[ 67%] Building CXX object code/CMakeFiles/code.dir/scripting/scripting.cpp.o
/usr/share/games/fs2source/fs2_open/code/scripting/scripting.cpp: In member function ‘bool ConditionedHook::ConditionsValid(int, object*, int)’:
/usr/share/games/fs2source/fs2_open/code/scripting/scripting.cpp:261:5: warning: this statement may fall through [-Wimplicit-fallthrough=]
     if(stricmp(Ship_types[sip->class_type].name, scp->data.name))
     ^~
/usr/share/games/fs2source/fs2_open/code/scripting/scripting.cpp:263:4: note: here
    case CHC_SHIPCLASS:
    ^~~~
```

```
[ 72%] Building CXX object code/CMakeFiles/code.dir/scripting/api/objs/order.cpp.o
/usr/share/games/fs2source/fs2_open/code/scripting/api/objs/order.cpp: In function ‘int scripting::api::l_Order_Target_f(lua_State*)’:
/usr/share/games/fs2source/fs2_open/code/scripting/api/objs/order.cpp:298:4: warning: this statement may fall through [-Wimplicit-fallthrough=]
    if (shipnum != -1){
    ^~
/usr/share/games/fs2source/fs2_open/code/scripting/api/objs/order.cpp:302:3: note: here
   case AI_GOAL_CHASE_WING:
   ^~~~
```

```
[ 83%] Building CXX object code/CMakeFiles/code.dir/stats/medals.cpp.o
/usr/share/games/fs2source/fs2_open/code/stats/medals.cpp: In function ‘void init_medal_bitmaps()’:
/usr/share/games/fs2source/fs2_open/code/stats/medals.cpp:771:6: warning: ‘%s’ directive writing up to 31 bytes into a region of size 30 [-Wformat-overflow=]
 void init_medal_bitmaps()
      ^~~~~~~~~~~~~~~~~~
/usr/share/games/fs2source/fs2_open/code/stats/medals.cpp:820:10: note: ‘sprintf’ output between 3 and 34 bytes into a destination of size 32
   sprintf(filename, "2_%s", Ranks[Player_score->rank].bitmap);
   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
[100%] Building CXX object freespace2/CMakeFiles/Freespace2.dir/freespace.cpp.o
/usr/share/games/fs2source/fs2_open/freespace2/freespace.cpp: In function ‘void game_leave_state(int, int)’:
/usr/share/games/fs2source/fs2_open/freespace2/freespace.cpp:5353:23: warning: this statement may fall through [-Wimplicit-fallthrough=]
    Training_num_lines = 0;
    ~~~~~~~~~~~~~~~~~~~^~~
/usr/share/games/fs2source/fs2_open/freespace2/freespace.cpp:5356:3: note: here
   case GS_STATE_GAME_PAUSED:
   ^~~~
```

### PR DESCRIPTION
**Pull request by asarium**
The two new warning types are switch-case fallthrough detection and `sprintf` truncation warnings. I fixed the first type by adding the fallthrough attribute where appropriate. Since `[[fallthrough]]` is a C++17 feature I added detection for other versions of the same attribute.

The second type was fixed by looking at the return value of `snprintf` and if the return value indicated a truncation then a null byte was written at the end of the buffer.

This fixes #1375.

### CODE REVIEW

File: code/parse/sexp.cpp [hunk 1]
```diff
@@ -26281,6 +26283,9 @@ int query_operator_argument_type(int op, int argnum)
 				return OPF_STRING;
 			} else if (argnum == 2) {
 				return OPF_VARIABLE_NAME;
+			} else {
+				// This shouldn't happen
```
Comment (MageKing17) [inline code/parse/sexp.cpp]: If it shouldn't happen, consider throwing an `Assertion(false)` in there.

File: code/parse/sexp.cpp [hunk 2]
```diff
@@ -26295,6 +26300,9 @@ int query_operator_argument_type(int op, int argnum)
 				return OPF_NUMBER;
 			} else if (argnum == 1) {
 				return OPF_VARIABLE_NAME;
+			} else {
+				// This shouldn't happen
```
Comment (MageKing17) [inline code/parse/sexp.cpp]: See previous comment.

File: code/parse/sexp.cpp [hunk 3]
```diff
@@ -26304,6 +26312,9 @@ int query_operator_argument_type(int op, int argnum)
 				return OPF_POSITIVE;
 			} else if (argnum == 3) {
 				return OPF_VARIABLE_NAME;
+			} else {
+				// This shouldn't happen
```
Comment (MageKing17) [inline code/parse/sexp.cpp]: And again.

File: code/parse/sexp.cpp [hunk 4]
```diff
@@ -26315,6 +26326,9 @@ int query_operator_argument_type(int op, int argnum)
 				return OPF_STRING;
 			} else if (argnum == 4) {
 				return OPF_VARIABLE_NAME;
+			} else {
+				// This shouldn't happen
```
Comment (MageKing17) [inline code/parse/sexp.cpp]: Ditto.

File: code/parse/sexp.cpp [hunk 5]
```diff
@@ -26723,6 +26737,9 @@ int query_operator_argument_type(int op, int argnum)
 				return OPF_SHIP;
 			else if (argnum == 1)
 				return OPF_SUBSYSTEM;
+			else
+				// This shouldn't happen
```
Comment (MageKing17) [inline code/parse/sexp.cpp]: Once more, with *feeling*.

File: code/parse/sexp.cpp [hunk 6]
```diff
@@ -26732,7 +26749,8 @@ int query_operator_argument_type(int op, int argnum)
 			else if (argnum == 2)
 				return OPF_SUBSYSTEM;
 			else
-				Int3();		// shouldn't happen
+				// This shouldn't happen
```
Comment (MageKing17) [inline code/parse/sexp.cpp]: Yeah, okay, I'm gonna stop doing individual comments for these.

File: code/parse/sexp.cpp (hunk 1 above)
Comment (asarium) [reply code/parse/sexp.cpp]: I'm pretty sure that that code should never be called with an invalid `argnum` value but if that happens then `OPF_NONE` is the correct return value (at least I think it's correct). I added the comment to signify that the `else` branch should never be called but if it is then it still returns a valid value.

**Pull request merged**